### PR TITLE
Add NodeGroup Name Into PVC Name

### LINKED ISSUE #4126: Support multiple LocalModel NodeGroup
/kind feature

**Describe the solution you'd like**
When you have multiple GPU types in the cluster like A100, H100:
* Models can be cached on multiple node groups, for each node group we create separate download jobs and localmodelnode controller updates with the expected model list.
* `LocalModelCache` can be updated to remove individual node group, e.g model is no longer needed to be cached on A100 nodes
* Different node groups may have different disk size, model size needs to be validated if the model can be cached or not for the  node groups.

```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: LocalModelCache
metadata:
  name: meta-llama3-8b-instruct
spec:
  modelSize: 10Gi
  nodeGroup: 
  - a100
  - h100
  sourceModelUri: hf://meta-llama/meta-llama-3-8b-instruct
```

when deploying `InferenceService`, node group annotation needs to be specified to schedule the workload to match the node group. `sourceModelUri` and `nodeGroup` need to be matched to use the local model cache.

```yaml
apiVersion: serving.kserve.io/v1beta1
    kind: InferenceService
    metadata:
      name: huggingface-llama3
      annotations:
        serving.kserve.io/nodegroup: a100
    spec:
      predictor:
        model:
          modelFormat:
            name: huggingface
          args:
            - --model_name=llama3
            - --model_id=meta-llama/meta-llama-3-8b-instruct
          resources:
            limits:
              cpu: "6"
              memory: 24Gi
              nvidia.com/gpu: "1"
            requests:
              cpu: "6"
              memory: 24Gi
              nvidia.com/gpu: "1"
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Links to the design documents:**
[Optional, start with the short-form [RFC template](https://docs.google.com/document/d/1UcBeLfZ_JMGpVrPJmYtEIVH_9Y4U3AEKQdq_IKuOMrU) to outline your ideas and get early feedback.]
[Required, use the longer-form [design doc template](https://docs.google.com/document/d/1Mtoui_PP2a9N59NjYHnsvrdJ8t2iKFwIJAx1zRO_I1c) to specify and discuss your design in more detail]

### PR DESCRIPTION
**Pull request by HotsauceLee**
**What this PR does / why we need it**: 
This is to make sure PVCs are unique for each node group and prepare supporting multiple node groups

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4126 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

**Review by greenmoon55 — APPROVED** [review]

**Review by greenmoon55 — COMMENTED** [review]
If an admin want to support different path for different node groups, they would have to create separate daemonsets on their own.

File: pkg/controller/v1alpha1/localmodel/controller.go [hunk 1]
```diff
@@ -201,9 +201,10 @@ func (c *LocalModelReconciler) ReconcileForIsvcs(ctx context.Context, localModel
 	}
 
 	for namespace := range namespaces {
+		pvcName := localModel.Name + "-" + localModel.Spec.NodeGroups[0]
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/localmodel/controller.go:205]: Later on when we support multiple node groups, the node group name needs to be retrieved from the inference service node group annotation

File: pkg/apis/serving/v1beta1/inference_service_defaults.go [hunk 2]
```diff
@@ -468,5 +468,6 @@ func (isvc *InferenceService) setLocalModelLabel(models *v1alpha1.LocalModelCach
 	}
 	isvc.Labels[constants.LocalModelLabel] = localModel.Name
 	isvc.Annotations[constants.LocalModelSourceUriAnnotationKey] = localModel.Spec.SourceModelUri
+	isvc.Annotations[constants.LocalModelPVCNameAnnotationKey] = localModel.Name + "-" + localModel.Spec.NodeGroups[0]
```
Comment (yuzisun) [inline pkg/apis/serving/v1beta1/inference_service_defaults.go:472]:
```suggestion
        // @TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
	isvc.Annotations[constants.LocalModelPVCNameAnnotationKey] = localModel.Name + "-" + localModel.Spec.NodeGroups[0]
```

File: pkg/controller/v1alpha1/localmodel/controller.go (hunk 1 above)
Comment (yuzisun) [inline pkg/controller/v1alpha1/localmodel/controller.go:205]:
```suggestion
                # @TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
		pvcName := localModel.Name + "-" + localModel.Spec.NodeGroups[0]
```

File: pkg/controller/v1alpha1/localmodel/controller.go [hunk 3]
```diff
@@ -201,9 +201,11 @@ func (c *LocalModelReconciler) ReconcileForIsvcs(ctx context.Context, localModel
 	}
 
 	for namespace := range namespaces {
+                # @TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/localmodel/controller.go]:
```suggestion
                // @TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```

File: pkg/apis/serving/v1beta1/inference_service_defaults.go [hunk 4]
```diff
@@ -468,5 +468,7 @@ func (isvc *InferenceService) setLocalModelLabel(models *v1alpha1.LocalModelCach
 	}
 	isvc.Labels[constants.LocalModelLabel] = localModel.Name
 	isvc.Annotations[constants.LocalModelSourceUriAnnotationKey] = localModel.Spec.SourceModelUri
+        // @TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```
Comment (yuzisun) [inline pkg/apis/serving/v1beta1/inference_service_defaults.go]:
```suggestion
        //@TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```

File: pkg/controller/v1alpha1/localmodel/controller.go [hunk 5]
```diff
@@ -201,9 +201,11 @@ func (c *LocalModelReconciler) ReconcileForIsvcs(ctx context.Context, localModel
 	}
 
 	for namespace := range namespaces {
+                // @TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/localmodel/controller.go]:
```suggestion
                //@TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```

File: pkg/apis/serving/v1beta1/inference_service_defaults.go [hunk 6]
```diff
@@ -468,5 +468,7 @@ func (isvc *InferenceService) setLocalModelLabel(models *v1alpha1.LocalModelCach
 	}
 	isvc.Labels[constants.LocalModelLabel] = localModel.Name
 	isvc.Annotations[constants.LocalModelSourceUriAnnotationKey] = localModel.Spec.SourceModelUri
+        //@TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```
Comment (yuzisun) [inline pkg/apis/serving/v1beta1/inference_service_defaults.go]:
```suggestion
        // TODO: node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```

File: pkg/controller/v1alpha1/localmodel/controller.go [hunk 7]
```diff
@@ -201,9 +201,11 @@ func (c *LocalModelReconciler) ReconcileForIsvcs(ctx context.Context, localModel
 	}
 
 	for namespace := range namespaces {
+                //@TODO node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/localmodel/controller.go]:
```suggestion
                // TODO: node group needs to be retrieved from isvc node group annotation when we support multiple node groups
```

**Pull request merged**